File: go/doubly_linked_list.go
package main

import (
	"encoding/json"
	"fmt"
)

type dllNode struct {
	value string
	prev  *dllNode
	next  *dllNode
}

// DoublyLinkedList представляет двусвязный список
type DoublyLinkedList struct {
	head *dllNode
	tail *dllNode
	size int
}

// NewDoublyLinkedList создает новый пустой двусвязный список
func NewDoublyLinkedList() *DoublyLinkedList {
	return &DoublyLinkedList{}
}

// PushFront добавляет элемент в начало списка
func (d *DoublyLinkedList) PushFront(value string) {
	newNode := &dllNode{value: value}

	if d.head == nil {
		d.head = newNode
		d.tail = newNode
	} else {
		newNode.next = d.head
		d.head.prev = newNode
		d.head = newNode
	}
	d.size++
}

// PushBack добавляет элемент в конец списка
func (d *DoublyLinkedList) PushBack(value string) {
	newNode := &dllNode{value: value}

	if d.tail == nil {
		d.head = newNode
		d.tail = newNode
	} else {
		newNode.prev = d.tail
		d.tail.next = newNode
		d.tail = newNode
	}
	d.size++
}

// PopFront удаляет и возвращает первый элемент
func (d *DoublyLinkedList) PopFront() (string, error) {
	if d.head == nil {
		return "", ErrEmptyContainer
	}

	value := d.head.value
	if d.head == d.tail {
		d.head = nil
		d.tail = nil
	} else {
		d.head = d.head.next
		d.head.prev = nil
	}
	d.size--
	return value, nil
}

// PopBack удаляет и возвращает последний элемент
func (d *DoublyLinkedList) PopBack() (string, error) {
	if d.tail == nil {
		return "", ErrEmptyContainer
	}

	value := d.tail.value
	if d.head == d.tail {
		d.head = nil
		d.tail = nil
	} else {
		d.tail = d.tail.prev
		d.tail.next = nil
	}
	d.size--
	return value, nil
}

// Insert вставляет элемент по указанному индексу
func (d *DoublyLinkedList) Insert(index int, value string) error {
	if index < 0 || index > d.size {
		return ErrIndexOutOfRange
	}

	if index == 0 {
		d.PushFront(value)
		return nil
	}

	if index == d.size {
		d.PushBack(value)
		return nil
	}

	var current *dllNode
	if index < d.size/2 {
		current = d.head
		for i := 0; i < index; i++ {
			current = current.next
		}
	} else {
		current = d.tail
		for i := d.size - 1; i > index; i-- {
			current = current.prev
		}
	}

	newNode := &dllNode{value: value, prev: current.prev, next: current}
	current.prev.next = newNode
	current.prev = newNode
	d.size++
	return nil
}

// Remove удаляет элемент по индексу
func (d *DoublyLinkedList) Remove(index int) error {
	if err := checkIndex(index, d.size); err != nil {
		return err
	}

	if index == 0 {
		_, err := d.PopFront()
		return err
	}

	if index == d.size-1 {
		_, err := d.PopBack()
		return err
	}

	var current *dllNode
	if index < d.size/2 {
		current = d.head
		for i := 0; i < index; i++ {
			current = current.next
		}
	} else {
		current = d.tail
		for i := d.size - 1; i > index; i-- {
			current = current.prev
		}
	}

	current.prev.next = current.next
	current.next.prev = current.prev
	d.size--
	return nil
}

// RemoveValue удаляет первое вхождение значения
func (d *DoublyLinkedList) RemoveValue(value string) bool {
	current := d.head
	for current != nil {
		if current.value == value {
			switch current {
			case d.head:
				_, err := d.PopFront()
				return err == nil
			case d.tail:
				_, err := d.PopBack()
				return err == nil
			default:
				current.prev.next = current.next
				current.next.prev = current.prev
				d.size--
				return true
			}
		}
		current = current.next
	}
	return false
}

// Get возвращает элемент по индексу
func (d *DoublyLinkedList) Get(index int) (string, error) {
	if err := checkIndex(index, d.size); err != nil {
		return "", err
	}

	if index < d.size/2 {
		current := d.head
		for i := 0; i < index; i++ {
			current = current.next
		}
		return current.value, nil
	}
	current := d.tail
	for i := d.size - 1; i > index; i-- {
		current = current.prev
	}
	return current.value, nil
}

// IndexOf возвращает индекс первого вхождения значения
func (d *DoublyLinkedList) IndexOf(value string) int {
	current := d.head
	index := 0
	for current != nil {
		if current.value == value {
			return index
		}
		current = current.next
		index++
	}
	return -1
}

// Contains проверяет наличие значения в списке
func (d *DoublyLinkedList) Contains(value string) bool {
	return d.IndexOf(value) != -1
}

// Size возвращает количество элементов
func (d *DoublyLinkedList) Size() int {
	return d.size
}

// IsEmpty проверяет пуст ли список
func (d *DoublyLinkedList) IsEmpty() bool {
	return d.size == 0
}

// Clear очищает список
func (d *DoublyLinkedList) Clear() {
	d.head = nil
	d.tail = nil
	d.size = 0
}

// Values возвращает все элементы списка
func (d *DoublyLinkedList) Values() []string {
	values := make([]string, 0, d.size)
	current := d.head
	for current != nil {
		values = append(values, current.value)
		current = current.next
	}
	return values
}

// Print выводит список в stdout
func (d *DoublyLinkedList) Print() {
	fmt.Print("DoublyLinkedList: nil <- ")
	current := d.head
	for current != nil {
		fmt.Printf("%s", current.value)
		if current.next != nil {
			fmt.Print(" <-> ")
		}
		current = current.next
	}
	fmt.Println(" -> nil")
}

// Serialize сериализует список в JSON
func (d *DoublyLinkedList) Serialize() ([]byte, error) {
	values := d.Values()
	return json.Marshal(values)
}

// Deserialize десериализует список из JSON
func (d *DoublyLinkedList) Deserialize(data []byte) error {
	var values []string
	if err := json.Unmarshal(data, &values); err != nil {
		return err
	}

	d.Clear()
	for _, value := range values {
		d.PushBack(value)
	}
	return nil
}

// SerializeBinary сериализует список в бинарный формат
func (d *DoublyLinkedList) SerializeBinary() ([]byte, error) {
	values := d.Values()
	type ListData struct {
		Size   int
		Values []string
	}

	listData := ListData{
		Size:   d.size,
		Values: values,
	}

	return serializeToBinary(listData)
}

// DeserializeBinary десериализует список из бинарного формата
func (d *DoublyLinkedList) DeserializeBinary(data []byte) error {
	type ListData struct {
		Size   int
		Values []string
	}

	var listData ListData
	if err := deserializeFromBinary(data, &listData); err != nil {
		return err
	}

	d.Clear()
	for _, value := range listData.Values {
		d.PushBack(value)
	}
	return nil
}

File: c++/doublylinkedlist.cpp
#include "doublylinkedlist.h"
#include <stdexcept>
#include <sstream>
#include <algorithm>

DoublyLinkedList::DoublyLinkedList() : head(nullptr), tail(nullptr), listSize(0) {}

DoublyLinkedList::~DoublyLinkedList() {
    clear();
}

DoublyLinkedList::DoublyLinkedList(const DoublyLinkedList& other) : head(nullptr), tail(nullptr), listSize(0) {
    Node* current = other.head;
    while (current != nullptr) {
        push_back(current->data);
        current = current->next;
    }
}

DoublyLinkedList& DoublyLinkedList::operator=(const DoublyLinkedList& other) {
    if (this != &other) {
        clear();
        Node* current = other.head;
        while (current != nullptr) {
            push_back(current->data);
            current = current->next;
        }
    }
    return *this;
}

void DoublyLinkedList::push_front(const std::string& value) {
    Node* newNode = new Node(value);
    
    if (empty()) {
        head = tail = newNode;
    } else {
        newNode->next = head;
        head->prev = newNode;
        head = newNode;
    }
    listSize++;
}

void DoublyLinkedList::push_back(const std::string& value) {
    Node* newNode = new Node(value);
    
    if (empty()) {
        head = tail = newNode;
    } else {
        tail->next = newNode;
        newNode->prev = tail;
        tail = newNode;
    }
    listSize++;
}

void DoublyLinkedList::pop_front() {
    if (empty()) {
        throw std::runtime_error("List is empty");
    }
    
    Node* temp = head;
    head = head->next;
    
    if (head != nullptr) {
        head->prev = nullptr;
    } else {
        tail = nullptr; // Список стал пустым
    }
    
    delete temp;
    listSize--;
}

void DoublyLinkedList::pop_back() {
    if (empty()) {
        throw std::runtime_error("List is empty");
    }
    
    Node* temp = tail;
    tail = tail->prev;
    
    if (tail != nullptr) {
        tail->next = nullptr;
    } else {
        head = nullptr; // Список стал пустым
    }
    
    delete temp;
    listSize--;
}

void DoublyLinkedList::insert(size_t index, const std::string& value) {
    if (index > listSize) {
        throw std::out_of_range("Index out of range");
    }
    
    if (index == 0) {
        push_front(value);
    } else if (index == listSize) {
        push_back(value);
    } else {
        Node* current = head;
        for (size_t i = 0; i < index; ++i) {
            current = current->next;
        }
        
        Node* newNode = new Node(value);
        newNode->prev = current->prev;
        newNode->next = current;
        current->prev->next = newNode;
        current->prev = newNode;
        
        listSize++;
    }
}

void DoublyLinkedList::remove(size_t index) {
    if (index >= listSize) {
        throw std::out_of_range("Index out of range");
    }
    
    if (index == 0) {
        pop_front();
    } else if (index == listSize - 1) {
        pop_back();
    } else {
        Node* current = head;
        for (size_t i = 0; i < index; ++i) {
            current = current->next;
        }
        
        current->prev->next = current->next;
        current->next->prev = current->prev;
        
        delete current;
        listSize--;
    }
}

std::string DoublyLinkedList::get(size_t index) const {
    if (index >= listSize) {
        throw std::out_of_range("Index out of range");
    }
    
    Node* current = head;
    for (size_t i = 0; i < index; ++i) {
        current = current->next;
    }
    return current->data;
}

size_t DoublyLinkedList::size() const {
    return listSize;
}

bool DoublyLinkedList::empty() const {
    return listSize == 0;
}

void DoublyLinkedList::clear() {
    while (head != nullptr) {
        Node* temp = head;
        head = head->next;
        delete temp;
    }
    tail = nullptr;
    listSize = 0;
}

void DoublyLinkedList::print() const {
    if (empty()) {
        std::cout << "DoublyLinkedList: [empty]" << std::endl;
        return;
    }
    
    Node* current = head;
    std::cout << "DoublyLinkedList: [";
    while (current != nullptr) {
        std::cout << "\"" << current->data << "\"";
        if (current->next != nullptr) std::cout << " <-> ";
        current = current->next;
    }
    std::cout << "]" << std::endl;
}

void DoublyLinkedList::serialize(std::ostream& os) const {
    os.write(reinterpret_cast<const char*>(&listSize), sizeof(listSize));
    Node* current = head;
    while (current != nullptr) {
        size_t strLen = current->data.length();
        os.write(reinterpret_cast<const char*>(&strLen), sizeof(strLen));
        os.write(current->data.c_str(), strLen);
        current = current->next;
    }
}

void DoublyLinkedList::deserialize(std::istream& is) {
    clear();
    size_t listSizeValue;
    is.read(reinterpret_cast<char*>(&listSizeValue), sizeof(listSizeValue));
    
    for (size_t i = 0; i < listSizeValue; ++i) {
        size_t strLen;
        is.read(reinterpret_cast<char*>(&strLen), sizeof(strLen));
        std::string value(strLen, ' ');
        is.read(&value[0], strLen);
        push_back(value);
    }
}

void DoublyLinkedList::serializeText(std::ostream& os) const {
    os << listSize << "\n";
    Node* current = head;
    while (current != nullptr) {
        // Экранируем специальные символы
        std::string escaped = current->data;
        size_t pos = 0;
        while ((pos = escaped.find('\n', pos)) != std::string::npos) {
            escaped.replace(pos, 1, "\\n");
            pos += 2;
        }
        pos = 0;
        while ((pos = escaped.find('\"', pos)) != std::string::npos) {
            escaped.replace(pos, 1, "\\\"");
            pos += 2;
        }
        os << "\"" << escaped << "\"\n";
        current = current->next;
    }
}

void DoublyLinkedList::deserializeText(std::istream& is) {
    clear();
    size_t newSize;
    is >> newSize;
    is.get(); // Пропускаем перевод строки
    
    for (size_t i = 0; i < newSize; ++i) {
        std::string line;
        std::getline(is, line);
        
        // Убираем обрамляющие кавычки и разэкранируем
        if (line.size() >= 2 && line.front() == '\"' && line.back() == '\"') {
            line = line.substr(1, line.size() - 2);
            
            // Разэкранируем специальные символы
            size_t pos = 0;
            while ((pos = line.find("\\n", pos)) != std::string::npos) {
                line.replace(pos, 2, "\n");
                pos += 1;
            }
            pos = 0;
            while ((pos = line.find("\\\"", pos)) != std::string::npos) {
                line.replace(pos, 2, "\"");
                pos += 1;
            }
        }
        
        push_back(line);
    }
}

// Новые методы для лучшего тестирования
bool DoublyLinkedList::checkIntegrity() const {
    if (empty()) {
        return head == nullptr && tail == nullptr;
    }
    
    // Проверка прямого прохода
    size_t forwardCount = 0;
    Node* current = head;
    const Node* last = nullptr;
    
    while (current != nullptr) {
        forwardCount++;
        last = current;
        current = current->next;
    }
    
    // Проверка обратного прохода
    size_t backwardCount = 0;
    current = tail;
    
    while (current != nullptr) {
        backwardCount++;
        current = current->prev;
    }
    
    return forwardCount == listSize && 
           backwardCount == listSize && 
           last == tail;
}

std::vector<std::string> DoublyLinkedList::toVector() const {
    std::vector<std::string> result;
    Node* current = head;
    while (current != nullptr) {
        result.push_back(current->data);
        current = current->next;
    }
    return result;
}
